src/dao/theatre.ts

```diff
@@ -0,0 +1,161 @@
+import db from "../utils/db";
+
+import { client, theatreCachePrefix } from "../cache";
+
+const redisKey = "allTheatres";
+class TheatreDAO {
+  async getAllTheatres() {
+    try {
+      const cachedTheatres = await client.get(redisKey);
+      if (cachedTheatres) {
+        return JSON.parse(cachedTheatres);
+      } else {
+        const theatres = await this.fetchTheatresFromDB();
+        await client.setEx(redisKey, 3600, JSON.stringify(theatres));
+        return theatres;
+      }
+    } catch (err) {
+      console.error(err);
+      return this.fetchTheatresFromDB();
+    }
+  }
+
+  private async fetchTheatresFromDB(): Promise<any[]> {
+    return new Promise((resolve, reject) => {
+      db.query("SELECT * FROM Theatre", (err, results) => {
+        if (err) {
+          reject(err);
+        } else {
+          resolve(results);
+        }
+      });
+    });
+  }
+
+  async getTheatreById(id: number) {
+    try {
+      const cachedTheatre = await client.get(`${theatreCachePrefix}${id}`);
+      if (cachedTheatre) {
+        return JSON.parse(cachedTheatre);
+      } else {
+        const theatre = await this.fetchTheatreByIdFromDB(id);
+        if (theatre) {
+          await client
+            .setEx(`${theatreCachePrefix}${id}`, 3600, JSON.stringify(theatre))
+            .catch((e) => console.error("Redis setEx error:", e));
+        }
+        return theatre;
+      }
+    } catch (err) {
+      console.error(err);
+      return this.fetchTheatreByIdFromDB(id);
+    }
+  }
+
+  private async fetchTheatreByIdFromDB(id: number): Promise<any> {
+    return new Promise((resolve, reject) => {
+      db.query(
+        "SELECT * FROM Theatre WHERE TheatreID = ?",
+        [id],
+        (err, results) => {
+          if (err) {
```

Comment:
provide a generic callback function at the places where you have followed this pattern